zookeeper-server/src/test/java/org/apache/zookeeper/test/BackupManagerTest.java

```diff
@@ -0,0 +1,601 @@
+        backupDir,
+        makeTmpName(BackupUtil.makeBackupName(Util.makeSnapshotName(1100), 1110)));
+    badFile.createNewFile();
+
+    badFile = new File(
+        backupDir,
+        makeTmpName(BackupUtil.makeBackupName(Util.makeLogName(1000), 1200)));
+    badFile.createNewFile();
+
+    backupManager.initialize();
+
+    actualBackupFiles = getBackupFiles(backupDir);
+
+    Assert.assertArrayEquals("bad files were should have been deleted",
+        expectedBackupFiles, actualBackupFiles);
+  }
+
+  private String makeTmpName(String filename) {
+    return "TMP_" + filename;
+  }
+
+  private void restoreAndValidate(int startingSnap, int endIndex, String[] nodeNames, boolean[] expectedExistence) throws Exception {
+    ZooKeeper c = null;
+
+    try {
+      c = closeRestoreRestartGetConnection(startingSnap, endIndex);
+
+      for (int n = 0; n < nodeNames.length && n < expectedExistence.length; n++) {
+        String msg = "Node " + nodeNames[n] + " existence after restore from snap "
+            + startingSnap + " and ending with log " + endIndex + ".";
+        Assert.assertEquals(msg, expectedExistence[n], exists(c, nodeNames[n]));
+        LOG.info("PASSED: " + msg);
+      }
+    } finally {
+      closeConnection(c);
+    }
+  }
+
+  private ZooKeeper closeRestoreRestartGetConnection(int startingSnap, int endIndex) throws Exception {
+    closeZk();
+    deleteDataFiles();
+    restoreFromBackup(startingSnap, endIndex);
+    startZk();
+    return createConnection();
+  }
+
+  private void deleteDataFiles() {
+    for (File f : snapLog.getSnapDir().listFiles()) {
+      LOG.info("Deleting file {}", f.getPath());
+      f.delete();
+    }
+    for (File f : snapLog.getDataDir().listFiles()) {
+      LOG.info("Deleting file {}", f.getPath());
+      f.delete();
+    }
+  }
+
+  private void restoreFromBackup(int startingSnap, int endPoint) throws Exception {
+    LOG.info("Restoring from backup: Snap {}  Log {}", startingSnap, endPoint);
+
+    List<BackupFileInfo> snaps = BackupUtil.getBackupFilesByMin(backupStorage, BackupFileType.SNAPSHOT);
+    List<BackupFileInfo> logs = BackupUtil.getBackupFilesByMin(backupStorage, BackupFileType.TXNLOG);
+
+    copyToLocal(snaps.get(startingSnap), snapLog.getSnapDir());
+
+    for (int i = startingSnap; i <= endPoint; i++) {
+      copyToLocal(logs.get(i), snapLog.getDataDir());
+    }
+  }
+
+  private void copyToLocal(BackupFileInfo src, File destDir) throws IOException {
+    String name = src.getStandardFile().getName();
+    File dest = new File(destDir, name);
+
+    LOG.info("Copying from {} to {}", src.getBackedUpFile(), dest);
+    backupStorage.copyToLocalStorage(src.getBackedUpFile(), dest);
+  }
+
+  private void setupZk() throws Exception {
+    LOG.info("Setting up directories");
+
+    dataDir = ClientBase.createTmpDir();
+    backupStatusDir = ClientBase.createTmpDir();
+    backupTmpDir = ClientBase.createTmpDir();
+    backupDir = ClientBase.createTmpDir();
```

Comment:
Same here, this dir path should match the one that FileSystemBackupStorage actually writes to.

Reply:
Fixed. Good catch!